primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,494 @@
+	}
+}
+
+/// A deadline based on block number and time.
+#[derive(Encode, Decode, Eq, PartialEq)]
+pub struct BlockAndTimeDeadline<B: BlockNumberProvider> {
+	/// The block number until which the lock is still valid _at least_.
+	pub block_number: <B as BlockNumberProvider>::BlockNumber,
+	/// The timestamp until which the lock is still valid _at least_.
+	pub timestamp: Timestamp,
+}
+
+impl<B: BlockNumberProvider> Clone for BlockAndTimeDeadline<B> {
+	fn clone(&self) -> Self {
+		Self {
+			block_number: self.block_number.clone(),
+			timestamp: self.timestamp.clone(),
+		}
+	}
+}
+
+impl<B: BlockNumberProvider> Default for BlockAndTimeDeadline<B> {
+	/// Provide the current state of block number and time.
+	fn default() -> Self {
+		Self {
+			block_number: B::current_block_number() + STORAGE_LOCK_DEFAULT_EXPIRY_BLOCKS.into(),
+			timestamp: offchain::timestamp().add(Duration::from_millis(
+				STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS,
+			)),
+		}
+	}
+}
+
+/// Lockable based on block number and timestamp.
+///
+/// Expiration is defined if both, block number _and_ timestamp
+/// expire.
+pub struct BlockAndTime<B: BlockNumberProvider> {
+	/// Relative block number offset, which is used to determine
+	/// the block number part of the deadline.
+	expiration_block_number_offset: u32,
+	/// Relative duration, used to derive the time based part of
+	/// the deadline.
+	expiration_duration: Duration,
+
+	#[doc(hidden)]
+	_phantom: core::marker::PhantomData<B>,
+}
+
+impl<B: BlockNumberProvider> Default for BlockAndTime<B> {
+	fn default() -> Self {
+		Self {
+			expiration_block_number_offset: STORAGE_LOCK_DEFAULT_EXPIRY_BLOCKS,
+			expiration_duration: Duration::from_millis(STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS),
+			_phantom: core::marker::PhantomData::<B>,
+		}
+	}
+}
+
+// derive not possible, since `B` does not necessarily implement `trait Clone`
+impl<B: BlockNumberProvider> Clone for BlockAndTime<B> {
+	fn clone(&self) -> Self {
+		Self {
+			expiration_block_number_offset: self.expiration_block_number_offset.clone(),
+			expiration_duration: self.expiration_duration,
+			_phantom: core::marker::PhantomData::<B>,
+		}
+	}
+}
+
+impl<B: BlockNumberProvider> Lockable for BlockAndTime<B> {
+	type Deadline = BlockAndTimeDeadline<B>;
+
+	fn deadline(&self) -> Self::Deadline {
+		let block_number = <B as BlockNumberProvider>::current_block_number()
+			+ self.expiration_block_number_offset.into();
+		BlockAndTimeDeadline {
+			timestamp: offchain::timestamp().add(self.expiration_duration),
+			block_number,
+		}
+	}
+
+	fn has_expired(deadline: &Self::Deadline) -> bool {
+		offchain::timestamp() > deadline.timestamp
+			&& <B as BlockNumberProvider>::current_block_number() > deadline.block_number
+	}
+
+	fn snooze(deadline: &Self::Deadline) {
+		let now = offchain::timestamp();
+		let remainder: Duration = now.diff(&(deadline.timestamp));
+		let snooze = core::cmp::min(remainder.millis(), STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE);
```

Comment:
In case the timeout has passed already and we are only limited by `block_number` deadline the `snooze` duration will always be `0` which will basically cause an endless busy loop.
It's probably fine, we can maybe at least skip calling `sleep_until` completely in such case, or if we want to be gentle we can introduce a minimal snooze time.